crates/accelerate/src/gate_direction.rs

```diff
@@ -97,7 +112,7 @@ where
 
         if let OperationRef::Instruction(py_inst) = packed_inst.op.view() {
             if py_inst.control_flow() {
-                let circuit_to_dag = imports::CIRCUIT_TO_DAG.get_bound(py); // TODO: Take out of the recursion
+                let circuit_to_dag = imports::CIRCUIT_TO_DAG.get_bound(py);
```

Comment:
This call here has two ways of being replaced:
    1. Using the `circuit_to_dag` rust function that would prevent you from extracting the Python `DAGCircuit`.
    2. Creating a parallel method that can do the recursion call with `QuantumCircuit` or `CircuitData` instances to skip the conversion entirely.

I'd honestly prefer option two since it doesn't seem that much is being done with the instructions that requires Python calls, and using `CircuitData::iter()` would allow you to iterate through the instructions in very much the same way.